src/redprl/operator.sml

```diff
@@ -347,34 +367,46 @@ struct
   in
     fun eqPoly f =
-      fn (LOOP r, LOOP r') => P.eq f (r, r')
-       | (HCOM (tag1, exs1, sp1), HCOM (tag2, exs2, sp2)) =>
-           tag1 = tag2
-             andalso extentsEq f (exs1, exs2)
-             andalso spanEq f (sp1, sp2)
-       | (COE (tag1, sp1), COE (tag2, sp2)) =>
-           tag1 = tag2 andalso spanEq f (sp1, sp2)
-       | (CUST (opid1, ps1, _), CUST (opid2, ps2, _)) =>
-           f (opid1, opid2) andalso paramsEq f (ps1, ps2)
-       | (RULE_LEMMA (opid1, ps1, _), RULE_LEMMA (opid2, ps2, _)) =>
-           f (opid1, opid2) andalso paramsEq f (ps1, ps2)
-       | (HYP_REF a, HYP_REF b) =>
-           f (a, b)
-       | (RULE_HYP (a, _), RULE_HYP (b, _)) =>
-           f (a, b)
-       | (RULE_ELIM (a, _), RULE_ELIM (b, _)) =>
-           f (a, b)
-       | (RULE_UNFOLD a, RULE_UNFOLD b) => 
-           f (a, b)
-       | (DEV_BOOL_ELIM a, DEV_BOOL_ELIM b) =>
-           f (a, b)
-       | (DEV_S1_ELIM a, DEV_S1_ELIM b) =>
-           f (a, b)
-       | (DEV_DFUN_ELIM a, DEV_DFUN_ELIM b) =>
-           f (a, b)
-       | (DEV_DPROD_ELIM a, DEV_DPROD_ELIM b) =>
-           f (a, b)
-       | _ => false
+      fn (LOOP r, t) => (case t of LOOP r' => P.eq f (r, r') | _ => false)
+       | (PATH_AP r, t) => (case t of PATH_AP r' => P.eq f (r, r') | _ => false)
+       | (HCOM (tag1, exs1, sp1), t) =>
+           (case t of
+                 HCOM (tag2, exs2, sp2) =>
+                   tag1 = tag2
+                   andalso extentsEq f (exs1, exs2)
+                   andalso spanEq f (sp1, sp2)
+               | _ => false)
+       | (COE (tag1, sp1), t) =>
+           (case t of
+                 COE (tag2, sp2) =>
+                   tag1 = tag2 andalso spanEq f (sp1, sp2)
+               | _ => false)
+       | (CUST (opid1, ps1, _), t) =>
+           (case t of
+                 CUST (opid2, ps2, _) =>
+                   f (opid1, opid2) andalso paramsEq f (ps1, ps2)
+               | _ => false)
+       | (RULE_LEMMA (opid1, ps1, _), t) =>
+           (case t of
+                 RULE_LEMMA (opid2, ps2, _) =>
+                   f (opid1, opid2) andalso paramsEq f (ps1, ps2)
+               | _ => false)
+       | (HYP_REF a, t) =>
+           (case t of HYP_REF b => f (a, b) | _ => false)
+       | (RULE_HYP (a, _), t) =>
+           (case t of RULE_HYP (b, _) => f (a, b) | _ => false)
+       | (RULE_ELIM (a, _), t) =>
+           (case t of RULE_ELIM (b, _) => f (a, b) | _ => false)
+       | (RULE_UNFOLD a, t) =>
+           (case t of RULE_UNFOLD b => f (a, b) | _ => false)
+       | (DEV_BOOL_ELIM a, t) =>
+           (case t of DEV_BOOL_ELIM b => f (a, b) | _ => false)
+       | (DEV_S1_ELIM a, t) =>
+           (case t of DEV_S1_ELIM b => f (a, b) | _ => false)
+       | (DEV_DFUN_ELIM a, t) =>
+           (case t of DEV_DFUN_ELIM b => f (a, b) | _ => false)
+       | (DEV_DPROD_ELIM a, t) =>
+           (case t of DEV_DPROD_ELIM b => f (a, b) | _ => false)
```

Comment:
Well, I am actually up for rewriting it back once the RedPRL core calculus is more or less stable (and that we have enough tests). Now, given that we already made a mistake, I prefer to stay with ugliness while we are still tempering with the heart of RedPRL.